src/libstd/fs.rs

```diff
@@ -1069,13 +1069,43 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
 }
 
 fn _remove_dir_all(path: &Path) -> io::Result<()> {
+    let filetype = try!(symlink_metadata(path)).file_type();
+    if filetype.is_symlink() {
+        if cfg!(windows) {
+            // On Windows symlinks to files and directories removed differently
+            // we should remove only directories here and have error on file
+            remove_dir(path)
+        } else {
+            // On unix symlinks are always files
+            remove_file(path)
+        }
```

Comment:
This might be good to extract into a helper function to be shared with below, something like:

```rust
fn remove(path: &Path, meta: &fs::Metadata) -> io::Result<()> {
    // ...
}
```